ecosystem/indexer-grpc/indexer-grpc-data-service-v2/src/connection_manager.rs

```diff
@@ -0,0 +1,291 @@
+    HistoricalDataServiceInfo, LiveDataServiceInfo, ServiceInfo, StreamInfo, StreamProgress,
+    StreamProgressSampleProto,
+};
+use dashmap::DashMap;
+use rand::prelude::*;
+use std::{
+    collections::VecDeque,
+    sync::atomic::{AtomicU64, Ordering},
+    time::{Duration, SystemTime},
+};
+use tonic::{codec::CompressionEncoding, transport::channel::Channel};
+use tracing::{info, warn};
+
+pub static MAX_HEARTBEAT_RETRIES: usize = 3;
+
+static OLD_PROGRESS_SAMPLING_RATE: Duration = Duration::from_secs(60);
+static RECENT_PROGRESS_SAMPLING_RATE: Duration = Duration::from_secs(1);
+static MAX_RECENT_SAMPLES_TO_KEEP: usize = 60;
+static MAX_OLD_SAMPLES_TO_KEEP: usize = 60;
+
+#[derive(Default)]
+struct StreamProgressSamples {
+    old_samples: VecDeque<StreamProgressSample>,
+    recent_samples: VecDeque<StreamProgressSample>,
+}
+
+impl StreamProgressSamples {
+    fn new() -> Self {
+        Default::default()
+    }
+
+    fn to_proto(&self) -> Vec<StreamProgressSampleProto> {
+        self.old_samples
+            .iter()
+            .chain(self.recent_samples.iter())
+            .map(|sample| StreamProgressSampleProto {
+                timestamp: Some(system_time_to_proto(sample.timestamp)),
+                version: sample.version,
+                size_bytes: sample.size_bytes,
+            })
+            .collect()
+    }
+
+    fn maybe_add_sample(&mut self, version: u64, size_bytes: u64) {
+        let now = SystemTime::now();
+        let sample = StreamProgressSample {
+            timestamp: now,
+            version,
+            size_bytes,
+        };
+
+        if Self::accept_sample(&self.recent_samples, &sample, RECENT_PROGRESS_SAMPLING_RATE) {
+            self.recent_samples.push_back(sample);
+            if self.recent_samples.len() > MAX_RECENT_SAMPLES_TO_KEEP {
+                let sample = self.recent_samples.pop_front().unwrap();
+                if Self::accept_sample(&self.old_samples, &sample, OLD_PROGRESS_SAMPLING_RATE) {
+                    self.old_samples.push_back(sample);
+                    if self.old_samples.len() > MAX_OLD_SAMPLES_TO_KEEP {
+                        self.old_samples.pop_front();
+                    }
+                }
+            }
+        }
+    }
+
+    fn accept_sample(
+        samples: &VecDeque<StreamProgressSample>,
+        sample: &StreamProgressSample,
+        sampling_rate: Duration,
+    ) -> bool {
+        if let Some(last_sample) = samples.back() {
+            if let Ok(delta) = sample.timestamp.duration_since(last_sample.timestamp) {
+                if delta >= sampling_rate {
+                    return true;
+                }
+            }
+        } else {
+            return true;
+        };
+
+        return false;
+    }
+}
+
+struct StreamProgressSample {
+    timestamp: SystemTime,
+    version: u64,
+    size_bytes: u64,
+}
+
+pub(crate) struct ConnectionManager {
```

Comment:
add some documentation comments here.

also i see that `String` is often used here as the service address. let's either use url/socketaddr or typedef it to avoid it mixed with regular strings.